hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java

```diff
@@ -856,23 +858,31 @@ private void validateSnapshotsExistAndActive(final String volumeName,
     checkSnapshotActive(toSnapInfo, false);
   }
 
-  public void validateSnapshotLimit() throws IOException {
+  /*
+   * Check snapshot limit
+   * Note: This method increments the snapshot in-flight counter
+   */
+  public synchronized void snapshotLimitCheck() throws IOException {
     OmMetadataManagerImpl omMetadataManager = (OmMetadataManagerImpl)
         ozoneManager.getMetadataManager();
     SnapshotChainManager snapshotChainManager =
         omMetadataManager.getSnapshotChainManager();
     int currentSnapshotNum = snapshotChainManager.getGlobalSnapshotChain().size();
-    if (currentSnapshotNum >= fsSnapshotMaxLimit) {
+    int inFlightCount = inFlightSnapshotCount.get();
```

Comment:
@peterxcli We should do this entire validation inside inFlightCount.update() block